File: leetcode/860.lemonade-change/go/lemonadechange.go
package lemonadechange

func lemonadeChange(bills []int) bool {
	tens := 0
	fives := 0

	for _, v := range bills {
		switch v {
		case 5:
			fives++
		case 10:
			{
				tens++
				fives--
			}
		case 20:
			{
				if tens > 0 {
					tens--
					fives--
				} else {
					fives -= 3
				}
			}
		}

		if tens < 0 || fives < 0 {
			return false
		}
	}

	return true
}

File: leetcode/860.lemonade-change/cpp/lemonadechange.cpp
#include <vector>

class Solution
{
public:
    bool lemonadeChange(std::vector<int> &bills)
    {
        int tens = 0;
        int fives = 0;

        for (auto v : bills)
        {
            switch (v)
            {
            case 5:
                fives++;
                break;
            case 10:
                tens++;
                fives--;
                break;
            case 20:
                tens--;
                fives--;
                if (tens < 0)
                {
                    tens++;
                    fives -= 2;
                }
                break;
            default:
                break;
            }

            if (fives < 0 || tens < 0)
                return false;
        }

        return true;
    }
};
